geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloIndexAdapter.scala

```diff
@@ -155,29 +156,71 @@ class AccumuloIndexAdapter(ds: AccumuloDataStore) extends IndexAdapter[AccumuloD
 
       case _ =>
         val (iter, eToF, reduce) = if (strategy.hints.isBinQuery) {
-          val iter = BinAggregatingIterator.configure(schema, index, ecql, hints)
-          (Seq(iter), new AccumuloBinResultsToFeatures(), None)
+          if (ds.config.remote.bin) {
+            val iter = BinAggregatingIterator.configure(schema, index, ecql, hints)
+            (Seq(iter), new AccumuloBinResultsToFeatures(), None)
+          } else {
+            if (hints.isSkipReduce) {
+              // override the return sft to reflect what we're actually returning,
+              // since the bin sft is only created in the local reduce step
+              hints.hints.put(QueryHints.Internal.RETURN_SFT, returnSchema)
+            }
+            (fti, resultsToFeatures, localReducer)
+          }
         } else if (strategy.hints.isArrowQuery) {
-          val (iter, reduce) = ArrowIterator.configure(schema, index, ds.stats, filter.filter, ecql, hints)
-          (Seq(iter), new AccumuloArrowResultsToFeatures(), Some(reduce))
+          if (ds.config.remote.arrow) {
+            val (iter, reduce) = ArrowIterator.configure(schema, index, ds.stats, filter.filter, ecql, hints)
+            (Seq(iter), new AccumuloArrowResultsToFeatures(), Some(reduce))
+          } else {
+            if (hints.isSkipReduce) {
+              // override the return sft to reflect what we're actually returning,
+              // since the arrow sft is only created in the local reduce step
+              hints.hints.put(QueryHints.Internal.RETURN_SFT, returnSchema)
+            }
+            (fti, resultsToFeatures, localReducer)
+          }
         } else if (strategy.hints.isDensityQuery) {
-          val iter = DensityIterator.configure(schema, index, ecql, hints)
-          (Seq(iter), new AccumuloDensityResultsToFeatures(), None)
+          if (ds.config.remote.density) {
+            val iter = DensityIterator.configure(schema, index, ecql, hints)
+            (Seq(iter), new AccumuloDensityResultsToFeatures(), None)
+          } else {
+            if (hints.isSkipReduce) {
+              // override the return sft to reflect what we're actually returning,
+              // since the density sft is only created in the local reduce step
+              hints.hints.put(QueryHints.Internal.RETURN_SFT, returnSchema)
+            }
+            (fti, resultsToFeatures, localReducer)
+          }
         } else if (strategy.hints.isStatsQuery) {
-          val iter = StatsIterator.configure(schema, index, ecql, hints)
-          val reduce = Some(StatsScan.StatsReducer(schema, hints))
-          (Seq(iter), new AccumuloStatsResultsToFeatures(), reduce)
+          if (ds.config.remote.stats) {
+            val iter = StatsIterator.configure(schema, index, ecql, hints)
+            val reduce = Some(StatsScan.StatsReducer(schema, hints))
+            (Seq(iter), new AccumuloStatsResultsToFeatures(), reduce)
+          } else {
+            if (hints.isSkipReduce) {
```

Comment:
Ok, I'll admit that I'm only half getting this. 

For a given 'function' (e.g. Density), are there situations where isSkipReduce is both true and false (when remote is false)?  If so, can you give an example of each?

Reply:
`skipReduce` will always be either true or false. It's used in the merged data store views, because if we run the reduce step we sometimes get things that can't be merged together easily (i.e. we go from arrow delta batches that we can merge to finalized arrow files that we can't). with remote encoding enabled, the intermediate (pre-reduce) features are already arrow/density/etc encoded, but if we do the encoding locally it's done in the reduce step. in that case, the simple feature type of the iterator doesn't match the simple feature type of the features in the iterator, which causes errors in the merge step when checking to see if the results were already encoded or not.

Reply:
Ok, I'm up on to 75% understanding.  For the last quarter, Does isSkipReduce==true when the features have been processed with their function remotely?

Reply:
when they're processed remotely, they get partially encoded in the remote iterator, so the return type is correct and doesn't need to be overridden.